File: problem_043.py
#!/usr/bin/env python
import re

print("Project Euler - Problem 43:")
print("Find the sum of all 0 to 9 pandigital numbers with the sub-string divisibility property.\n")

def permute(num):
    length = len(num)
    used = [0, 0, 0, 0, 0, 0, 0, 0, 0, 0]       # Array to store numbers in!
    used[int(num[length-1])] = 1                # Prime array
    for i in range(length-2, -1, -1):           # Start at right and move left
        used[int(num[i])] += 1                  # Keep track of digits encountered
        if int(num[i]) < int(num[i+1]):         # Look for first digit that is smaller than digit to right
            newNum = str(num[:i])               # New number is everything to left of digit encountered

            for n in range(int(num[i])+1, 10):  # Look for next biggest number in used array
                if used[n] > 0:                 # If array contains an available number
                    used[n] -= 1                # Take one out
                    newNum += str(n)            # Append it to our new number string
                    break                       # Get out of here!

            for n in range(0, 10):              # Now go through the used array (smallest first)
                while used[n] > 0:              # If we have available digits
                    used[n] -= 1                # Take out out
                    newNum += str(n)            # Append it to out new number string

            return newNum                       # Return the number

def subStringy(n):
    primes = [2, 3, 5, 7, 11, 13, 17]
    for i in range(7):
        if int(n[1+i:i+4]) % primes[i] != 0: return False
    return True

n = "1406357289"
sum = 0
while n[0] < "9":
    if subStringy(n) == True:
        sum += int(n)
    n = permute(n)

print(sum)

File: problem_043.js
#!/usr/bin/node

console.log("Project Euler - Problem 43:");
console.log("Find the sum of all 0 to 9 pandigital numbers with the sub-string divisibility property.\n");

// Find all 2-3 digit numbers divisible by the first 7 primes
var prime_sub_numbers = {
    2  : [],
    3  : [],
    5  : [],
    7  : [],
    11 : [],
    13 : [],
    17 : [],
};

for (var n=1; n<1000; n++){
    for (p in prime_sub_numbers)
        if (n*p < 1000 && n*p > 9)
            prime_sub_numbers[p].push(n*p);
}

// Use recursion to construct these numbers with sub-string divisibility
var potential_pandigitals = [];
function build_number(n, s){
    if (s.length == 0){
        potential_pandigitals.push(n);
        return n;
    }

    for (var p of prime_sub_numbers[s[0]]){
        if (Math.floor(n % 100) == Math.floor(p / 10)){
            build_number(n * 10 + Math.floor(p % 10), s.slice(1));
        }
    }
}

for (var n of prime_sub_numbers[2]){
    build_number(n, [3, 5, 7, 11, 13, 17]);
}

// Sort through the potential pandigitals and add up only the real pandigitals
function is_pandigital(s){
    var digits = [0, 0, 0, 0, 0, 0, 0, 0, 0, 0];

    // Count up digits
    for (c of s.toString())
        digits[Number(c)]++;

    // One digit should be missing. This will be the leading digit
    // If we have more than one missing digit or two of anything return 0
    var missing = -1;
    for (var d of digits){
        if (d == 0){
            if (missing == -1){
                missing = d;
            } else {
                return 0;
            }
        }
        if (d > 1)
            return 0;
    }

    return Number(missing + s);
}

var sum = 0;
for (var pp of potential_pandigitals)
    sum += is_pandigital(pp);

console.log(sum);
